### PR TITLE
Hot module reload with `react-refresh`

### PR DESCRIPTION
**Pull request by akiselev**
This PR enables refresh-free hot module reload for React components using [`react-refresh`](https://github.com/facebook/react/tree/main/packages/react-refresh) which is apparently the officially supported React HMR library.

### Repro

https://user-images.githubusercontent.com/653604/166645859-1b73a7ec-d05a-4904-bcca-6bdf2dd58304.mp4

<br/>

<br/>

## Known Caveats

I copy-pasted the below from the [`react-refresh-webpack-plugin` troubleshooting docs](https://github.com/pmmmwh/react-refresh-webpack-plugin/blob/main/docs/TROUBLESHOOTING.md). Much of our codebase violates the below caveats (especially Class components and unnamed/pascal named Components) so state isn't preserved, though it shouldn't force a full reload most of the time.

### State not preserved for class components

This is intended behaviour.
If you're coming from [react-hot-loader](https://github.com/gaearon/react-hot-loader), this might be the biggest surprise for you.

The main rationale behind this is listed in [this wish list for hot reloading](https://overreacted.io/my-wishlist-for-hot-reloading/):

> It is better to lose local state than to behave incorrectly.
>
> It is better to lose local state than use an old version.

The truth is, historically, hot reloading for class components was never reliable and maintainable.
It was implemented based on workarounds like wrapping components with Proxies.

While these workarounds made hot reloading "work" on the surface, they led to inconsistencies in other departments:

- Lifecycle methods will fire randomly at random times;
- Type checks will fail randomly (`<Component />.type === Component` is `false`); and
- Mutation of React's internals, which is difficult to manage and will need to play catch up with React as we move into the future (a-la Concurrent Mode).

Thus, to keep fast refresh reliable and resilient to errors (with recovery for most cases), class components will always be re-mounted on a hot update.

### Edits always lead to full reload

In most cases, if the plugin is applied correctly, it would mean that we weren't able to set up boundaries to stop update propagation.
It can be narrowed down to a few unsupported patterns:

1. Un-named/non-pascal-case-named components

   See [this tweet](https://twitter.com/dan_abramov/status/1255229440860262400) for drawbacks of not giving component proper names.
   They are impossible to support because we have no ways to statically determine they are React-related.
   This issue also exist for other React developer tools, like the [hooks ESLint plugin](https://www.npmjs.com/package/eslint-plugin-react-hooks).
   Internal components in HOCs also have to conform to this rule.

   ```jsx
   // won't work
   export default () => <div />;
   export default function () {
     return <div />;
   }
   export default function divContainer() {
     return <div />;
   }
   ```

2. Chain of files leading to root with none containing React-related content only

   This pattern cannot be supported because we cannot ensure non-React-related content are free of side effects.
   Usually with this error you will see something like this in the browser console:

   ```
   Ignored an update to unaccepted module ... [a very long path]
   ```

3. `export * from 'namespace'` (TypeScript only)

   This only affect users using TypeScript on Babel.
   This pattern is only supported when you don't mix normal exports with type exports, or when all your exports conform to the `PascalCase` rule.
   This is because we cannot statically analyse the exports from the namespace to determine whether we can set up a boundary and stop update propagation.

### CODE REVIEW

**Review by gusaiani — APPROVED** [review]
Tested locally, works well for me, a lot smoother in many many situations. 👏👏👏👏

**Review by losrebellos — APPROVED** [review]

**Comment by losrebellos** [comment]
Seems to be all good 👏

File: frontend/src/metabase/dev-noop.js [hunk 1]
```diff
@@ -0,0 +1,2 @@
+// Make sure this gets recognized as an ES module
+export const noop = () => {};
```
Comment (WiNloSt) [inline frontend/src/metabase/dev-noop.js:2]: I think `export {}` might work. And probably be better than just exporting something we never used. Could you try that out?

File: webpack.config.js [hunk 2]
```diff
@@ -221,26 +224,30 @@ if (WEBPACK_BUNDLE === "hot") {
...
-      // because newer versions would require us to upgrade to react-router v4 and possibly deal with
-      // asynchronous route issues as well. See https://github.com/gaearon/react-hot-loader/issues/249
-      { loader: "react-hot-loader/webpack" },
-      { loader: "babel-loader", options: BABEL_CONFIG },
+      {
+        loader: "babel-loader",
+        options: {
+          ...BABEL_CONFIG,
+          plugins: [
+            "@emotion",
```
Comment (ranquild) [inline webpack.config.js:235]: Do we need to repeat plugins here? I thought it was unnecessary.

File: frontend/src/metabase/dev-noop.js (hunk 1 above)
Comment (ranquild) [reply frontend/src/metabase/dev-noop.js:2]: You could also use this file which is already used for EE plugins https://github.com/metabase/metabase/blob/4953ab8575a9acbdfe999349e0ecc39e38700663/frontend/src/metabase/lib/noop.js#L5-L4

File: webpack.config.js (hunk 2 above)
Comment (akiselev) [reply webpack.config.js:235]: This overrides `.babelrc` plugins to add `react-refresh/babel`

File: frontend/src/metabase/dev-noop.js (hunk 1 above)
Comment (akiselev) [reply frontend/src/metabase/dev-noop.js]: @ranquild using `metabase/lib/noop.js` results in the error:

```
[js] ERROR in ./app-main.js 5:0-22
[js] Module not found: Error: Can't resolve 'metabase-dev' in '/home/kiselev/git/metabase/metabase/frontend/src/metabase'
```

@WiNloSt nice catch! updated

**Pull request merged**